### PR TITLE
DB-2117  separate table service to use another nginx ingress controller

### PR DESCRIPTION
**Pull request by rluidash**
Changes in this PR 

Background, this change is associated to ticket [DB-2117 ](https://dashbase.atlassian.net/browse/DB-2117) in which Dialpad want to whitelist table endpoints; but unable to do it. This is because the default ingress "ingress" include web,auth, and table services. Whitelisting or put FW rules on table  endpoint public IP will also affect web and auth service because they are sharing or using the same ingress and nginx ingress controller.

one of the workaround solution is to separate table and other services; in which table services have their own ingress e.g. "ingress-table" and  that using separate nginx ingress controller "nginx-ingress-table".
image

Details of the solution is described in https://docs.google.com/document/d/1V-lv3KyXmngd0I8CJVbUaDTQnwliHcO_Xtnh3pqShlE/edit?usp=sharing

For the solution, it requires change in dashbase helm chart ; and the  ingress.yaml  file need to  to have condition for table services to opt out in default ingress "ingress" and use separate ingress "ingress-table"  . And all these changes can be done via updating the custom value file dashbase-values.yaml.

And by default, table expose via ingress together with other services e.g. web,auth and grafana. The ingress section in the dashbase-values.yaml file is kept the same as original. This will not affect current customers

```
ingress:
  enabled: true
  certmanager: false
  host: test.dashbase.io
```

On the other hand, customer or user if planned to use separate nginx ingress controller e.g. "nginx-ingress-table" which is configured with ingress.class =  nginx-table ; can modify the dashbas-values.yaml file like below

```
ingress:
  enabled: true
  certmanager: false
  includetable: false
  ingresstable: true
  host: test.dashbase.io
```

- The includetable key is specified false means not include table service in default ingress “ingress” . And by default this value is true, so it won't affect current setup.
- The ingresstable key is specified “true” means create a new  ingress “ingress-table” and include table service within it. And the ingress “ingress-table” has annotation for ingress.class using “nginx-table”

The changes in files for this PR are below
ingress.yaml 
values.yaml

And dashbase helm chart won't automatically create separate nginx ingress controller.  User can do that manually using command below or use dashbase installer script.

`helm install nginx-ingress-table stable/nginx-ingress --set controller.ingressClass=nginx-table --set controller.useIngressClassOnly=true --namespace dashbase`

### CODE REVIEW

**Comment by rluidash** [comment]
Tested this change with following cases
V1 (1.5.4)  install with default ingress 
V1 (1.5.4)  helm upgrade  to include default ingress + ingress-table
V1 (1.5.4)  install with default ingress +  ingress-table

V1 (2.1.0-rc1)  install with default ingress 
V1 (2.1.0-rc1)  helm upgrade  to include default ingress + ingress-table
V1 (2.1.0-rc1)  install with default ingress +  ingress-table

And all results are positive

File: dashbase/templates/ingress.yaml [hunk 1]
```diff
@@ -38,12 +38,16 @@ spec:
         {{- end }}
       {{- end }}
       {{- range $name, $_ := .Values.tables }}
+        {{- if $.Values.ingress.includetable }}
     - {{ printf "table-%v.%v" $name $host }}
+        {{- end }}
       {{- end }}
       {{- range $name, $val := .Values.tablesv2 }}
-      {{- if $val.indexer }}
+        {{- if $.Values.ingress.includetable }}
```
Comment (sincejune) [inline dashbase/templates/ingress.yaml:46]: this can be simplified in one single if condition.

File: dashbase/templates/ingress.yaml [hunk 2]
```diff
@@ -62,21 +66,25 @@ spec:
...
     http:
       paths:
        - backend:
            serviceName: table-{{ $name }}
            servicePort: 7888
   {{- end }}
+  {{- end }}
   {{- range $name, $val := .Values.tablesv2 }}
-  {{- if $val.indexer }}
+  {{- if $.Values.ingress.includetable }}
```
Comment (sincejune) [inline dashbase/templates/ingress.yaml:79]: ditto

File: dashbase/templates/ingress.yaml [hunk 3]
```diff
@@ -128,3 +136,69 @@ spec:
...
+{{- $host := (.Values.ingress.host | default (printf "%s.k8s" .Release.Name)) -}}
+kind: Ingress
+apiVersion: extensions/v1beta1
+metadata:
+  name: ingress-table
+  labels:
+    app: {{.Release.Name}}
+    component: ingress
+  annotations:
+    kubernetes.io/ingress.class: {{ .Values.ingress.className | default "nginx-table"}}
```
Comment (sincejune) [inline dashbase/templates/ingress.yaml:151]: shall we use `.Values.ingress.tableClassName` or something else?

File: dashbase/templates/ingress.yaml (hunk 1 above)
Comment (rluidash) [reply dashbase/templates/ingress.yaml]: I simplify that into one single condition

File: dashbase/templates/ingress.yaml (hunk 2 above)
Comment (rluidash) [reply dashbase/templates/ingress.yaml]: I simplify that into one single condition

File: dashbase/templates/ingress.yaml (hunk 3 above)
Comment (rluidash) [reply dashbase/templates/ingress.yaml]: yes, changed  to `.Values.ingress.tableClassName`

**Comment by rluidash** [comment]
I rerun all tests after Chao's comment change.

Tested this change with following cases
V1 (1.5.4) install with default ingress
V1 (1.5.4) helm upgrade to include default ingress + ingress-table
V1 (1.5.4) install with default ingress + ingress-table

V1 (2.2.0) install with default ingress
V1 (2.2.0) helm upgrade to include default ingress + ingress-table
V1 (2.2.0) install with default ingress + ingress-table

And all results are positive

**Comment by rluidash** [comment]
currently to toggle to use  separate ingress  for table endpoints require.. two flags under ingress section of dashbase-values.yaml file..

- includetable
- ingresstable

```
ingress:
  enabled: true
  certmanager: false
  includetable: false
  ingresstable: true
  host: raytest14.dashbase.io
```
 With that setup.. you can expose tables  in both ingress and ingress-table ..  if you have both of these flags be true

  includetable: true
  ingresstable: true

I will leave now, since need more time to analysis the effect but will create an improvement ticket to combine with possibility options to group other service endpoints in another ingress controller(s).

**Pull request merged**